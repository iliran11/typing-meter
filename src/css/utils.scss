@import 'css/colors.scss';

.row {
  display: flex;
}
.full-width {
  width: 100%;
}
.background-primary {
  background-color: $primary-color;
}
.text-center {
  text-align: center;
}
.flex-center {
  display: flex;
  align-items:center;
}
