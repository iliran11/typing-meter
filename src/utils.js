import randomWords from 'random-words';
import { WORDS_AMOUNT } from './constants';
import keyIndex from './react-key-index';
import scrollIntoView from 'scroll-into-view';
// import ow from 'ow';

export function generateLoremIpsum() {
  return Array.from(new Array(WORDS_AMOUNT), () => {
    return randomWords();
  });
}
export function processTextToArray(text) {
  // ow(text, ow.string);
  return replaceLineBreaks(text)
    .split(' ')
    .filter(filterEmptyStrings);
}
export function createIndexWordObjects(wordsArray) {
  const indexedWordsArray = keyIndex(wordsArray, getRandomNumber());
  return indexedWordsArray.map(word => {
    return createWordObject({ challenge: word.value, key: word.id });
  });
}
export function secondstoMillisecond(number) {
  return number * 1000;
}
export function millisecondsToSeconds(number) {
  const seconds = number / 1000;
  return Math.ceil(seconds);
}
export function millisecondsToMinutes(number) {
  return number / 60000;
}
export function createWordObject({ challenge = '', typed = '', key }) {
  return {
    challenge,
    typed,
    key,
    get isCompleted() {
      const { challenge, typed } = this;
      return challenge.length <= typed.length;
    },
    get isEmpty() {
      const { typed } = this;
      const trimmedTyped = typed.trim();
      return trimmedTyped.length === 0;
    },
    get isCorrect() {
      const { challenge, typed } = this;
      const relevantTyped = typed.substr(0, challenge.length).toLowerCase();
      return this.ChallengeLowerCase === relevantTyped;
    },
    get ChallengeLowerCase() {
      return this.challenge.toLowerCase();
    },
    get wordArray() {
      return this.challenge.split('');
    },
    get typedArray() {
      return this.typed.split('');
    },
    get challenegeArray() {
      return this.challenge.split('');
    },
    get challengeLength() {
      return this.challenegeArray.length;
    },
    get removeLastTypedLetter() {
      return this.typed.slice(0, -1);
    },

    get numberOfCorrectEntities() {
      return this.typedArray.reduce(
        (accumulator, currentValue, currentIndex) => {
          const isLetterCorrect = currentValue === this.wordArray[currentIndex];
          if (isLetterCorrect) accumulator.correct += 1;
          else {
            accumulator.wrong += 1;
          }
          return accumulator;
        },
        { correct: 0, wrong: 0 }
      );
    }
  };
}
function filterEmptyStrings(value) {
  return value !== '';
}
export function replaceLineBreaks(string) {
  /**https://stackoverflow.com/questions/784539/how-do-i-replace-all-line-breaks-in-a-string-with-br-tags?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa */
  return string.replace(/(?:\r\n|\r|\n)/g, ' ');
}
export function isLastCharIsSpace(str) {
  // https://stackoverflow.com/questions/3884632/how-to-get-the-last-character-of-a-string?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa  return str.substr(str.length - 1, str.length - 2) === ' ';
  return str.slice(-1) === ' ';
}
export const noop = () => {};

export function getRandomNumber() {
  return Math.floor(Math.random() * 200) + 1;
}

/** focusing the next word needed to be focus */
export function focusNode(node) {
  const wordsContainer = node.parentNode.parentNode;
  /** height of the words container */
  const containerHeight = wordsContainer.clientHeight;
  /** amount from the upper point of the scorlling window, to the top of the actual element */
  const containerScorllingOffset = wordsContainer.scrollTop;
  /** number of pixels from the top of the element to the center of the scrolling window */
  const threshold = containerHeight * 0.5 + containerScorllingOffset;
  /** check if the next active word is not aligned with threshold */
  const isNextWordOutsideThreshold = node.offsetTop !== threshold;
  /** if the word is below or above the thresehold - make it higher */
  if (isNextWordOutsideThreshold) {
    scrollIntoView(
      node,
      {
        time: 200,
        align: {
          top: 0.2
        },
        isScrollable: () => true
      },
      () => {}
    );
  }
}
