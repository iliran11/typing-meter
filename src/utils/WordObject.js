export default function createWordObject({ challenge = '', typed = '', key }) {
  return {
    challenge,
    typed,
    key,
    get isCompleted() {
      const { challenge, typed } = this;
      return challenge.length <= typed.length;
    },
    get isEmpty() {
      const { typed } = this;
      const trimmedTyped = typed.trim();
      return trimmedTyped.length === 0;
    },
    get isCorrect() {
      const { challenge, typed } = this;
      const relevantTyped = typed.substr(0, challenge.length).toLowerCase();
      return this.ChallengeLowerCase === relevantTyped;
    },
    get ChallengeLowerCase() {
      return this.challenge.toLowerCase();
    },
    get wordArray() {
      return this.challenge.split('');
    },
    get typedArray() {
      return this.typed.split('');
    },
    get challenegeArray() {
      return this.challenge.split('');
    },
    get challengeLength() {
      return this.challenegeArray.length;
    },
    get removeLastTypedLetter() {
      return this.typed.slice(0, -1);
    },

    get numberOfCorrectEntities() {
      return this.typedArray.reduce(
        (accumulator, currentValue, currentIndex) => {
          const isLetterCorrect = currentValue === this.wordArray[currentIndex];
          if (isLetterCorrect) accumulator.correct += 1;
          else {
            accumulator.wrong += 1;
          }
          return accumulator;
        },
        { correct: 0, wrong: 0 }
      );
    }
  };
}
