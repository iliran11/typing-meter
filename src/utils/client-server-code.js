import {
  createGame,
  createRandomWordsArray,
  updateWordNextStatus
} from './gameUtils';

import * as constants from '../constants';
import { wpmScore } from './wpmCalculations';
export default {
  updateWordNextStatus,
  createRandomWordsArray,
  constants,
  wpmScore,
  createGame,
};
