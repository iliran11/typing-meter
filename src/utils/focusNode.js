import scrollIntoView from 'scroll-into-view';

/** focusing the next word needed to be focus */
export default function focusNode(node) {
  const wordsContainer = node.parentNode.parentNode;
  /** height of the words container */
  const containerHeight = wordsContainer.clientHeight;
  /** amount from the upper point of the scorlling window, to the top of the actual element */
  const containerScorllingOffset = wordsContainer.scrollTop;
  /** number of pixels from the top of the element to the center of the scrolling window */
  const threshold = containerHeight * 0.5 + containerScorllingOffset;
  /** check if the next active word is not aligned with threshold */
  const isNextWordOutsideThreshold = node.offsetTop !== threshold;
  /** if the word is below or above the thresehold - make it higher */
  if (isNextWordOutsideThreshold) {
    scrollIntoView(
      node,
      {
        time: 200,
        align: {
          top: 0.2
        },
        isScrollable: () => true
      },
      () => {}
    );
  }
}