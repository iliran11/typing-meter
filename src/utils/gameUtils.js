import uuid from 'uuid';
import isNull from 'lodash.isnull';
import {WPM_NULL, AWAITS_TYPING, GAME_DURATION } from '../constants';
import {
  padWordsWithSpaces,
  secondstoMillisecond,
  createIndexWordObjects,
  getRandomNumber,
  generateLoremIpsum,
  processTextToArray
} from './utils';

function wordsArray(customWordsState) {
  if (isNull(customWordsState)) {
    return generateLoremIpsum();
  }
  return processTextToArray(this.props.customWords);
}

export function createGame() {
  const gameId = `game-${uuid()}`;
  const overallTime = secondstoMillisecond(GAME_DURATION);
  const wordsArrayPaddedWithSpaces = padWordsWithSpaces(wordsArray(null));
  const words = createIndexWordObjects(
    wordsArrayPaddedWithSpaces,
    getRandomNumber()
  );
  return {
    overallTime,
    gameId,
    index: 0,
    scrollIndex: 0,
    words,
    wpm: WPM_NULL,
    gameDuration: GAME_DURATION,
    gameStatus: AWAITS_TYPING
  };
}
