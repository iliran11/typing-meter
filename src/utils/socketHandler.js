import io from 'socket.io-client';
import {
  CREATE_GAME,
  CREATE_COMPETITOR_GAME,
  PLAYER_JOINED_GAME,
  YOU_JOINED_ROOM,
  COMPETITOR_JOINED_GAME
} from '../constants';
import { createMyGame, addCompetitor } from '../store/games/gameActions';

const socketHandler = {
  initSocket(dispatch) {
    this.socket = io('http://localhost:4000');
    this.socket.on(YOU_JOINED_ROOM, data => {
      createMyGame(
        { words: data.words, gameId: data.gameId, players: data.players },
        dispatch
      );
    });
    this.socket.on(COMPETITOR_JOINED_GAME,playerObject=>{
      addCompetitor(playerObject,dispatch)
    })
  },
  emitEvent(eventName, data) {
    this.socket.emit(eventName, data);
  }
};
export default socketHandler;
