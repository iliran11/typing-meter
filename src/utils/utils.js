import randomWords from 'random-words';
import { WORDS_AMOUNT } from '../constants';
import keyIndex from '../react-key-index';
import createWordObject from './WordObject';
// import ow from 'ow';

export function generateLoremIpsum() {
  return Array.from(new Array(WORDS_AMOUNT), () => {
    return randomWords();
  });
}
export function processTextToArray(text) {
  // ow(text, ow.string);
  return replaceLineBreaks(text)
    .split(' ')
    .filter(filterEmptyStrings);
}
export function createIndexWordObjects(wordsArray, key) {
  const indexedWordsArray = keyIndex(wordsArray, key);
  return indexedWordsArray.map(word => {
    return createWordObject({ challenge: word.value, key: word.id });
  });
}
export function secondstoMillisecond(number) {
  return number * 1000;
}
export function millisecondsToSeconds(number) {
  const seconds = number / 1000;
  return Math.ceil(seconds);
}

export function filterEmptyStrings(value) {
  return value !== '';
}
export function replaceLineBreaks(string) {
  /**https://stackoverflow.com/questions/784539/how-do-i-replace-all-line-breaks-in-a-string-with-br-tags?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa */
  return string.replace(/(?:\r\n|\r|\n)/g, ' ');
}
export function isLastCharIsSpace(str) {
  // https://stackoverflow.com/questions/3884632/how-to-get-the-last-character-of-a-string?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa  return str.substr(str.length - 1, str.length - 2) === ' ';
  return str.slice(-1) === ' ';
}
export function getLastCharInString(str) {
  return str.substr(str.length - 1)
}
export const noop = () => {};

export function getRandomNumber() {
  return Math.floor(Math.random() * 200) + 1;
}

export function padWordsWithSpaces(words) {
  return words.reduce((accumulator, word, currentIndex) => {
    const isLastIndex = currentIndex + 1 === words.length;
    if (isLastIndex) return accumulator;
    return [...accumulator, word, ' '];
  }, []);
}
