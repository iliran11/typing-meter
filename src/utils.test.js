import {processTextToArray} from './utils'

describe('process custom text to array', () => {
  test('3 normal words', () => {
    const textSample = `hello bye ffff`
    expect(processTextToArray(textSample)).toMatchSnapshot();
  });
  test('empty string', () => {
    const textSample = ``
    expect(processTextToArray(textSample)).toMatchSnapshot();
  });
  test('empty string with spaces', () => {
    const textSample = `       `
    expect(processTextToArray(textSample)).toMatchSnapshot();
  });
  test('single letter word', () => {
    const textSample = `hello . hello`
    expect(processTextToArray(textSample)).toMatchSnapshot();
  });
  test('padding text', () => {
    const textSample = `     hello hello     `
    expect(processTextToArray(textSample)).toMatchSnapshot();
  });
  test('line breaks', () => {
    const textSample = `hello
       bye 
       
       line breaks`
    expect(processTextToArray(textSample)).toMatchSnapshot();
  });

})