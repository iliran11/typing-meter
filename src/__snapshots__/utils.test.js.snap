// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`process custom text to array 3 normal words 1`] = `
Array [
  "hello",
  "bye",
  "ffff",
]
`;

exports[`process custom text to array empty string 1`] = `Array []`;

exports[`process custom text to array empty string with spaces 1`] = `Array []`;

exports[`process custom text to array line breaks 1`] = `
Array [
  "hello",
  "bye",
  "line",
  "breaks",
]
`;

exports[`process custom text to array padding text 1`] = `
Array [
  "hello",
  "hello",
]
`;

exports[`process custom text to array single letter word 1`] = `
Array [
  "hello",
  ".",
  "hello",
]
`;
